### ISSUE TITLE
Save and restore problem

### ISSUE THREAD

**Issue by xisnu** [igormq/ctc_tensorflow_example#11]
Thank you for this excellent example. I tried to implement my model which is similar to this. But I have been facing a problem during saving and restoring the model. When the training is going on everything is fine and weights are saved. But when I am stopping the iterations and then load the model with saved weights then CTC loss is showing some previous values and not the last one where it has been saved. Any help will be highly appreciated.

**Comment by igormq** [comment]
Hi, @xisnu ! Could you provide your code?

**Comment by xisnu** [comment]
Thanks for your interest @igormq 
This is the part where I am creating the computation graph
```
graph=tf.Graph()
with graph.as_default():
    print("Graph Creation")
    x=tf.placeholder(tf.float32,[None,ms,nb_features],name="x")

    y=tf.sparse_placeholder(tf.int32,name="y")
    seq_len=tf.placeholder(tf.int32,[None],name="seq_len")
    

    f_cell=tf.nn.rnn_cell.LSTMCell(nb_hidden,state_is_tuple=True)
    f_stack=tf.nn.rnn_cell.MultiRNNCell([f_cell]*nb_layers)
    
    b_cell=tf.nn.rnn_cell.LSTMCell(nb_hidden,state_is_tuple=True)
    b_stack=tf.nn.rnn_cell.MultiRNNCell([b_cell]*nb_layers)
    
    outputs,_=tf.nn.bidirectional_dynamic_rnn(f_stack,b_stack,x,sequence_length=seq_len,dtype=tf.float32)

    merge=tf.concat(2, outputs,name="merge")

    shape = tf.shape(x)
    batch_s,maxtimesteps=shape[0],shape[1]

    output_reshape = tf.reshape(merge, [-1, nb_hidden*2])#batch*timesteps,nb_hidden
   
    W = tf.Variable(tf.truncated_normal([nb_hidden*2,nb_classes],stddev=0.1),name="W1")

    b = tf.Variable(tf.constant(0., shape=[nb_classes]),name="b1")

    logits = tf.add(tf.matmul(output_reshape, W) , b,name="logits") #818622,52
      
    logits_reshape = tf.transpose(tf.reshape(logits, [batch_s, -1, nb_classes]),[1,0,2],name="logits_reshape")#534,1533,52

    loss =tf.nn.ctc_loss(logits_reshape, y, seq_len,time_major=True)
    cost = tf.reduce_mean(loss,name="cost")

    optimizer = tf.train.RMSPropOptimizer(lr).minimize(cost)

    decoded, log_prob = tf.nn.ctc_greedy_decoder(logits_reshape, seq_len)

    actual_ed=tf.edit_distance(tf.cast(decoded[0], tf.int32),y,normalize=False)
    ler = tf.reduce_sum(actual_ed,name="ler")
    saver=tf.train.Saver() #for storing last state
    bestsaver=tf.train.Saver() #for saving best state
    print("Network Ready")
```
Now I am running the session according to
```
with tf.Session(graph=graph) as session:    
    #saver = tf.train.Saver()
    if(sys.argv[1]=="load"):
        saver.restore(session, "Weights/model_last")
        print("Previous weights loaded")
    else:
        init_op = tf.global_variables_initializer()
        session.run(init_op)
        print("New Weights Initialized")
    
    best=0
    acctest=0
    bestloss=10000
   
    for e in range(nb_epochs):
        f=open(logfilename,"a")
        totalloss=0
        totalacc=0
        starttime=time.time()

        for b in range(trainbatch):
            p=b+1
            print("Reading Batch ",p,"/",trainbatch,end="\r")
            feed = {x:inpx[b],y:inp_sparse_y[b],seq_len:inpseqlen[b]}#These arrays are batches
            batchloss,batchacc, _ = session.run([cost,ler,optimizer], feed)
            
            totalloss=totalloss+batchloss
            totalacc=totalacc+batchacc
        
        avgloss=totalloss/trainbatch
        avgacc=1-(totalacc/nctr)
        saver.save(session, "Weights/model_last")
        
        testloss=0
        testacc=0
        for t in range(testbatch):
            testfeed = {x:test_inpx[t],y:test_inp_sparse_y[t],seq_len:test_inpseqlen[t]}
            outcome,testbatchloss,testbatchacc=session.run([decoded[0],cost,ler],testfeed)
            if(t==0):
                first_batch_outcome=outcome
            testloss=testloss+testbatchloss
            testacc=testacc+testbatchacc
        
        testfile=open("Results.txt","w")
        testfile.write("Epoch "+str(e)+"\n")
        for tc in range(len(testcases)):
            predicted=label_from_sparse(first_batch_outcome,testcases[tc])
            testfile.write(str(true[tc])+" As "+str(predicted)+"\n")
        testfile.close()
        
        testloss=testloss/testbatch
        testacc=1-(testacc/ncts)
        
        endtime=time.time()        
        if(testacc>best):
            best=testacc
            print("Test Result Improvement")
            bestsaver.save(session, "BestWeights/model_best")
        timetaken=endtime-starttime
        msg="Epoch "+str(e)+"("+str(timetaken)+ " sec ) Training: Cost is "+str(avgloss)+" Accuracy "+str(avgacc)+" Testing: Loss "+str(testloss)+" Accuracy "+str(testacc)+"\n"
        print(msg)
        f.write(msg)
        f.close()
```
I am loading my data (online handwriting) prior to the creation of graph. The training is running fine. Accuracy is increasing upto 81%. But whenever I am stopping the model and loading it again it is showing some previous loss and accuracy value instead of last saved one.

**Comment by abrari** [comment]
I'm using this to initialize the Saver object and it saves / restores well

```
saver = tf.train.Saver(tf.global_variables())
```

**Comment by igormq** [comment]
Thanks @abrari , I think the issue was solved. Closing.
